### PR TITLE
Proof of concept of overwritable fields

### PR DESCRIPTION
**Pull request by jszwedko**
I have an unusual use-case where I'd like to overwrite fields previously added to a logger via `With(rz.Fields())` to support recursive `context.Context` usage.

Specifically, I've created a tracer for the https://github.com/99designs/gqlgen library where I add the current field being resolved to the logger (stored and retrieved from the `context.Context`), but, as GraphQL fields are resolved recursively, I end up with JSON having duplicate keys in children fields like:

```json
{
  "gql_resolver_object":"Query",
  "gql_resolver_field":"me",
  "gql_resolver_alias":"me",
  "resolver_path": ["me"],

  "gql_resolver_object":"User",
  "gql_resolver_field":"email",
  "gql_resolver_alias":"email",
  "resolver_path":["me", "email"],

  "event":"gql_field_resolved",
  "duration":149846132,
  "gql_errors": [],
  "timestamp":"2019-07-24T14:23:02Z",
  "message":"gql field resolved"
}
```

My formatting added for clarity.

What I'd like is for the last field added with a given key to be applied when writing out the log message.

I took a stab at adding this functionality here via an `Overwritable` function implementing the `Field` signature that, instead of sending it directly to the encoder, merely stores it in a `map`. When the event is written out, the fields in that map are then resolved. This allows a user to opt into this behavior for fields they know that they will be overwriting as I think this likely adds some overhead to the logging.

Example code using my added function:

```golang
package main

import "github.com/bloom42/rz-go"

func main() {
	l := rz.New()

	l = l.With(
		rz.Fields(
			rz.Overwritable("component", rz.String("component", "foo")),
		),
	)
	l = l.With(
		rz.Fields(
			rz.Overwritable("component", rz.String("component", "bar")),
		),
	)

	l.Info("hello")
}
```

Which results in:

```json
{"level":"info","component":"bar","timestamp":"2019-07-24T14:29:53Z","message":"hello"}
```

Instead of:

```json
{"level":"info","component":"foo","component":"bar","timestamp":"2019-07-24T15:00:35Z","message":"hello"}
```

I think I'm probably missing some edge cases, but wanted to get this up to see if this approach is acceptable. If it is, I'll polish it and add tests.

Thank you!

### CODE REVIEW

File: logger.go [hunk 1]
```diff
@@ -178,6 +179,9 @@ func (l *Logger) logEvent(level LogLevel, message string, done func(string), fie
 	for i := range fields {
 		fields[i](e)
 	}
+	for i := range e.overwritableFields {
```
Comment (z0mbie42) [inline logger.go]: because `e.overwritableFields` is a map, I would prefer `for key := range e.overwritableFields {`, becasue `i` make me think that we are iterating an index based slice

**Comment by z0mbie42** [comment]
Hi,
Thank you for the work!

Yes this approach is totally  acceptable!

The double key thing (`rz.Overwritable("component", rz.String("component"`) is a little bit disturbing, if we do

```go
	l = l.With(
		rz.Fields(
			rz.Overwritable("MyIndexMap", rz.String("component", "foo")),
		),
	)
	l = l.With(
		rz.Fields(
			rz.Overwritable("MyIndexMap", rz.Int("other_omponent", 42)),
		),
	)
```
then  we will have something like
```
{"level":"info","other_omponent":42,"timestamp":"2019-07-24T14:29:53Z","message":"hello"}
```

right?

Does it still fit you use case if we find a way to  implement it like:
```go
	l = l.With(
		rz.Fields(
			rz.Overwritable("component", "foo"),
		),
	)
	l = l.With(
		rz.Fields(
			rz.Overwritable("component", 42),
		),
	)
```
?

**Comment by jszwedko** [comment]
@z0mbie42 thanks for reviewing!

I agree that the duplicating the key to also pass to `rz.Overwritable` can lead to some confusion, but I was unsure how else to implement it than adding an `OverwritableX` for each existing `Field`, for example `rz.OverwritableInt8`, to be able to capture the key while maintaining type safety.

I suppose we could have just one `rz.Overwritable` that matches the behavior of `rz.Any` and uses reflection. Is that what you are thinking?

**Comment by jszwedko** [comment]
I have considered one other approach that might solve the key issue which would be:

* Have `rz.Overwritable` simply take an `rz.Field` without a key
* When outputting the event, serialize overwritable fields to a temporary event
* Deserialize the temporary event and depend on `json.Unmarshal`s behavior of overwriting keys when deserializing into a `map[string]interface{}`to ensure we get the last one for a given key.

This would look like:

```diff
diff --git a/event.go b/event.go
index 3dca094..dce212b 100644
--- a/event.go
+++ b/event.go
@@ -38,7 +38,7 @@ type Event struct {
 	formatter            LogFormatter
 	timestampFunc        func() time.Time
 	encoder              Encoder
-	overwritableFields   map[string]Field
+	overwritableFields   []Field
 }
 
 func putEvent(e *Event) {
@@ -90,11 +90,8 @@ func (e *Event) Append(fields ...Field) {
 	}
 }
 
-func (e *Event) overwritable(key string, field Field) {
-	if e.overwritableFields == nil {
-		e.overwritableFields = make(map[string]Field)
-	}
-	e.overwritableFields[key] = field
+func (e *Event) overwritable(field Field) {
+	e.overwritableFields = append(e.overwritableFields, field)
 }
 
 // Fields returns the fields from the event.
diff --git a/field.go b/field.go
index 52e580a..6268f60 100644
--- a/field.go
+++ b/field.go
@@ -59,9 +59,9 @@ func String(key, value string) Field {
 // Adding another Overwritable field with the same key will cause the previous one to be dropped.
 //
 // Typically the `key` will match the `key` passed to the method creating `Field`, but does not have to.
-func Overwritable(key string, field Field) Field {
+func Overwritable(field Field) Field {
 	return func(e *Event) {
-		e.overwritable(key, field)
+		e.overwritable(field)
 	}
 }
 
diff --git a/logger.go b/logger.go
index 95f4374..a30b656 100644
--- a/logger.go
+++ b/logger.go
@@ -37,7 +37,7 @@ type Logger struct {
 	timestampFunc        func() time.Time
 	contextMutext        *sync.Mutex
 	encoder              Encoder
-	overwritableFields   map[string]Field
+	overwritableFields   []Field
 }
 
 // New creates a root logger with given options. If the output writer implements
@@ -179,9 +179,14 @@ func (l *Logger) logEvent(level LogLevel, message string, done func(string), fie
 	for i := range fields {
 		fields[i](e)
 	}
-	for key := range e.overwritableFields {
-		e.overwritableFields[key](e)
+
+	overwritableEvent := newEvent(l.writer, level)
+	copyInternalLoggerFieldsToEvent(l, overwritableEvent)
+	for i := range e.overwritableFields {
+		overwritableEvent.overwritableFields[i](overwritableEvent)
 	}
+	overwritableFields, _ := overwritableEvent.Fields()
+	e.fields(overwritableFields)
 
 	writeEvent(e, message, done)
 }
```

I assume that deserializing the JSON would make overwritable fields much more expensive but does allow us to preserve the type safety of the `rz.Field` functions.

What do you think of that approach vs. having `rz.Overwritable` behave like `rz.Any`? I personally lean towards it behaving like `rz.Any` instead.

**Comment by jszwedko** [comment]
@z0mbie42 just circling back to this. I'm curious if you have a preference for ether approach:

* temporary JSON marshaling and unmarshaling
* having a method signature match `rz.Any`
* creating a matching `OverwritableX` for each existing `rz.Field` implementation (for example. `rz.OverwritableInt`)

I'm happy to implement any of them, but have a preference for `rz.Any` as it meets our use case and depending on `json.Unmarshal` overwriting keys in `map[string]interface{}` seems a little like depending on undefined behavior.

**Comment by z0mbie42** [comment]
Hi,
I apologize for ignoring this PR for such long time, but I'm focus on my current work (https://gitlab.com/bloom42). Also we no longer use Go (in favor of Rust) so we no longer use `rz-go`. All of that make that I do not have the bandwidth to maintain the repository, and wrap my head around how to improve things/how changes can go wrong.

Thus I'm closing this PR without merging, feel free to fork.

Sincerely,
Sylvain Kerkour
